### PR TITLE
SAMZA-1220 : Add thread name to SamzaContainer shutdown hook and prevent shutdown deadlock

### PR DESCRIPTION
**Pull request by navina**
* SamzaContainerExceptionHandler is written in Java and used by LocalContainerRunner.java

### CODE REVIEW

**Comment by navina** [comment]
@prateekm Can you please take a look at the changes? Please comment if this was not what you had in mind. Thanks!

File: samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala [hunk 1]
```diff
@@ -803,21 +806,40 @@ class SamzaContainer(
...
         }
-        runLoopThread.join(shutdownMs)
-        if (runLoopThread.isAlive) {
-          warn("Did not shut down within %s ms, exiting" format shutdownMs)
-        } else {
-          info("Shutdown complete")
+        try {
+          runLoopThread.join(shutdownMs)
+        } catch {
+          case e: Exception => // Ignore to avoid deadlock with uncaughtExceptioHandler. See SAMZA-1220
```
Comment (vjagadish1989) [inline samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala]: IMHO it makes sense to catch, log and ignore all `Throwable` here.

- OOM errors happening in shutdown hook are usually in-actionable.
- We will bound the wait time by `shutdownMs` anyways.
- Other throwables happening in the `shutdownHook` will end up calling the `UncaughtExceptionHandler` resulting in a potential deadlock. Catching all `Throwable`s will prevent this.

File: samza-core/src/main/java/org/apache/samza/container/SamzaContainerExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+package org.apache.samza.container;
+
+import java.lang.Thread.UncaughtExceptionHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An UncaughtExceptionHandler for SamzaContainer that simply shuts down when any thread throws
+ * an uncaught exception.
```
Comment (navina) [inline samza-core/src/main/java/org/apache/samza/container/SamzaContainerExceptionHandler.java]: I think this is a result of copying the documentation over from the scala class. I will fix the documentation.

File: samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala [hunk 3]
```diff
@@ -801,21 +804,40 @@ class SamzaContainer(
...
+  }
+
+  def removeShutdownHook = {
+    try {
+      if (shutdownHookThread != null) {
+        Runtime.getRuntime.removeShutdownHook(shutdownHookThread)
+      }
+    } catch {
+      case e: IllegalStateException => {
+        // When samza is shutdown by external command, IllegalStateException will be thrown.
```
Comment (prateekm) [inline samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala]: Maybe rephrase to:
"Thrown when then JVM is already shutting down, so safe to ignore."?

File: samza-core/src/test/java/org/apache/samza/container/TestSamzaContainerExceptionHandler.java [hunk 4]
```diff
@@ -17,20 +17,23 @@
...
-package org.apache.samza.container
+package org.apache.samza.container;
 
-import org.junit.Test
-import org.junit.Assert._
-import org.junit.Before
-import org.apache.samza.SamzaException
-import org.junit.After
+import org.apache.samza.SamzaException;
+import org.junit.Test;
```
Comment (navina) [inline samza-core/src/test/java/org/apache/samza/container/TestSamzaContainerExceptionHandler.java]: That is how IntelliJ organizes imports. Were you referring some specific ordering?

File: samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala (hunk 1 above)
Comment (navina) [inline samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala]: Sure. Makes sense.

File: samza-core/src/main/java/org/apache/samza/container/SamzaContainerExceptionHandler.java (hunk 2 above)
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/container/SamzaContainerExceptionHandler.java]: +1 if you want to remove the Runnable parameter.

File: samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala (hunk 1 above)
Comment (prateekm) [inline samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala]: +1 for catching Throwable. Maybe also move the logging to try instead of finally (in case it throws).

File: samza-core/src/test/java/org/apache/samza/container/TestSamzaContainerExceptionHandler.java (hunk 4 above)
Comment (vjagadish1989) [inline samza-core/src/test/java/org/apache/samza/container/TestSamzaContainerExceptionHandler.java]: Makes sense, I thought the `junit` imports could have been potentially grouped together. I think I may have missed the `static` clause in the `assertTrue`.

File: samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala (hunk 3 above)
Comment (navina) [inline samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala]: Ok.

**Review by vjagadish1989 — APPROVED** [review]
approved. looks great to me. thank you for the contribution navina.
